core/encoding.go

```diff
@@ -245,51 +244,33 @@ func DecodeIndexDataStoreKey(
 
 		i := len(key.Fields)
 		descending := false
+		var kind client.FieldKind = client.FieldKind_DocID
 		// If the key has more values encoded then fields on the index description, the last
 		// value must be the docID and we treat it as a string.
+		if i < len(fields) {
+			kind = fields[i].Kind
+		}
 		if i < len(indexDesc.Fields) {
 			descending = indexDesc.Fields[i].Descending
 		} else if i > len(indexDesc.Fields) {
 			return IndexDataStoreKey{}, ErrInvalidKey
 		}
 
-		var val any
-		data, val, err = encoding.DecodeFieldValue(data, descending)
+		var val client.NormalValue
+		data, val, err = encoding.DecodeFieldValue(data, descending, kind)
 		if err != nil {
 			return IndexDataStoreKey{}, err
 		}
+		if i == len(indexDesc.Fields) {
+			if _, ok := val.String(); !ok {
+				return IndexDataStoreKey{}, NewErrInvalidFieldValue("docID is not a string")
```

Comment:
todo: Please document how/when this can happen, or remove it.  I am struggling to imagine how this might happen (besides manually messing around in the datastore).

Reply:
removed